From: NeilBrown <neilb@suse.de>
Subject: nfs_prime_dcache needs fh to be set.
References: bnc#908069 bnc#896484
PAtch-mainline: Submitted, http://www.spinics.net/lists/linux-nfs/msg49753.html


In some circumstances, READDIRPLUS does not return a filehandle.
A particular example is that the NFSv3 Linux server does not return a
filehandle for directories that are mounted on.

When that happens, we mustn't call nfs_prime_cache, as it can
mess up the mount table.

This is clearly a bug, and this is a safe file.  It is not certain if
it addresses either the referenced bugzilla issues.  It might.

Signed-off-by: NeilBrown <neilb@suse.de>

---
 fs/nfs/dir.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/fs/nfs/dir.c
+++ b/fs/nfs/dir.c
@@ -546,7 +546,7 @@ int nfs_readdir_page_filler(nfs_readdir_
 
 		count++;
 
-		if (desc->plus != 0)
+		if (desc->plus != 0 && entry->fh->size > 0)
 			nfs_prime_dcache(desc->file->f_path.dentry, entry);
 
 		status = nfs_readdir_add_to_array(entry, page);
